File: SourceCppRun/SynEditDelphiInstances.pas
unit SynEditDelphiInstances;

interface

{$HPPEMIT '#pragma link "SynEditDelphiInstances"'}

uses
  System.Generics.Collections, System.Generics.Defaults, Vcl.Controls;

type

  // A single fold
  // Important: FromLine, ToLine are 1-based
  TSynFoldRange = record
    FromLine: Integer; // Beginning line
    ToLine: Integer; // End line
    FoldType: Integer;  // Could be used by some complex highlighters
    Indent : Integer;   // Only used for Indent based folding (Python)
    Collapsed: Boolean; // Is collapsed?
  private
    function GetLinesCollapsed: Integer;
  public
    procedure Move(Count: Integer);
    property LinesCollapsed: Integer read GetLinesCollapsed;
    constructor Create(AFromLine : Integer; AToLine : Integer = -1;
      AFoldType : Integer = 1; AIndent : Integer = -1;
      ACollapsed : Boolean = False);
  end;

  PSynFoldRange = ^TSynFoldRange;

  TFoldOpenClose = (focOpen, focClose);

  TLineFoldInfo = record
    Line : Integer;
    FoldOpenClose : TFoldOpenClose;
    FoldType : Integer;
    Indent : Integer;
    constructor Create(ALine : Integer;
      AFoldOpenClose : TFoldOpenClose = focOpen;
      AFoldType : Integer = 1; AIndent : Integer = -1);
  end;

  TBufferCoord = record
    Char: integer;
    Line: integer;
    class operator Equal(a, b: TBufferCoord): Boolean;
    class operator NotEqual(a, b: TBufferCoord): Boolean;
    class operator LessThan(a, b: TBufferCoord): Boolean;
    class operator LessThanOrEqual(a, b: TBufferCoord): Boolean;
    class operator GreaterThan(a, b: TBufferCoord): Boolean;
    class operator GreaterThanOrEqual(a, b: TBufferCoord): Boolean;
    class function Min(a, b: TBufferCoord): TBufferCoord; static;
    class function Max(a, b: TBufferCoord): TBufferCoord; static;
  end;

  TSynUndoItem = class abstract(TObject)
    ChangeNumber: Integer;
    FCaret: TBufferCoord;
    GroupBreak: Boolean;
  public
    procedure Undo(Editor: TCustomControl); virtual; abstract;
    procedure Redo(Editor: TCustomControl); virtual; abstract;
  end;


TSynFoldRangesCmp = class(TInterfacedObject, System.Generics.Defaults.IComparer<TSynFoldRange>)
public
  function Compare(const Left, Right: TSynFoldRange): Integer;
end;

TLineFoldInfoCmp = class(TInterfacedObject, System.Generics.Defaults.IComparer<TLineFoldInfo>)
public
  function Compare(const Left, Right: TLineFoldInfo): Integer;
end;

function CreateFoldRangeList: System.Generics.Collections.TList<TSynFoldRange>;
function CreateFoldInfoList: System.Generics.Collections.TList<TLineFoldInfo>;

procedure AddToFoldRangeList(AList: System.Generics.Collections.TList<TSynFoldRange>; ARange: TSynFoldRange);
function BinarySearchInFoldRangeList(AList: System.Generics.Collections.TList<TSynFoldRange>; ARange: TSynFoldRange; var AIndex: Integer): Boolean;

// for SynEditCodeFolding
procedure InsertIntoLineFoldList(AList: System.Generics.Collections.TList<TLineFoldInfo>; AIndex: Integer; AInfo: TLineFoldInfo);
function BinarySearchInLineFoldList(AList: System.Generics.Collections.TList<TLineFoldInfo>; AInfo: TLineFoldInfo; var AIndex: Integer): Boolean;

implementation


function TSynFoldRangesCmp.Compare(const Left, Right: TSynFoldRange): Integer;
begin
  Result := Left.FromLine - Right.FromLine;
end;

function TLineFoldInfoCmp.Compare(const Left, Right: TLineFoldInfo): Integer;
begin
  Result := Left.Line - Right.Line;
end;


function CreateFoldRangeList: System.Generics.Collections.TList<TSynFoldRange>;
begin
  result := System.Generics.Collections.TList<TSynFoldRange>.Create(TSynFoldRangesCmp.Create);
end;

function CreateFoldInfoList: System.Generics.Collections.TList<TLineFoldInfo>;
begin
  result := System.Generics.Collections.TList<TLineFoldInfo>.Create(TLineFoldInfoCmp.Create);
end;


{ TSynEditFoldRange }

constructor TSynFoldRange.Create(AFromLine, AToLine, AFoldType: Integer;
  AIndent : Integer; ACollapsed: Boolean);
begin
  FromLine := AFromLine;
  ToLine := AToLine;
  FoldType := AFoldType;
  Indent := AIndent;
  Collapsed := ACollapsed;
end;

function TSynFoldRange.GetLinesCollapsed: Integer;
begin
  if Collapsed then
    Result := ToLine - FromLine
  else
    Result := 0;
end;

procedure TSynFoldRange.Move(Count: Integer);
begin
  Inc(FromLine, Count);
  Inc(ToLine, Count);
end;

constructor TLineFoldInfo.Create(ALine: Integer;
  AFoldOpenClose: TFoldOpenClose; AFoldType: Integer; AIndent : Integer);
begin
    Line := ALine;
    FoldOpenClose := AFoldOpenClose;
    FoldType := AFoldType;
    Indent := AIndent;
end;

procedure AddToFoldRangeList(AList: System.Generics.Collections.TList<TSynFoldRange>; ARange: TSynFoldRange);
begin
  AList.Add(ARange);
end;

function BinarySearchInFoldRangeList(AList: System.Generics.Collections.TList<TSynFoldRange>; ARange: TSynFoldRange; var AIndex: Integer): Boolean;
begin
  result := AList.BinarySearch(ARange, AIndex);
end;

// for SynEditCodeFolding
procedure InsertIntoLineFoldList(AList: System.Generics.Collections.TList<TLineFoldInfo>; AIndex: Integer; AInfo: TLineFoldInfo);
begin
  AList.Insert(AIndex, AInfo);
end;

function BinarySearchInLineFoldList(AList: System.Generics.Collections.TList<TLineFoldInfo>; AInfo: TLineFoldInfo; var AIndex: Integer): Boolean;
begin
  result := AList.BinarySearch(AInfo, AIndex);
end;

{ TBufferCoord }

class operator TBufferCoord.Equal(a, b: TBufferCoord): Boolean;
begin
  Result := (a.Char = b.Char) and (a.Line = b.Line);
end;

class operator TBufferCoord.GreaterThan(a, b: TBufferCoord): Boolean;
begin
  Result :=  (b.Line < a.Line)
    or ((b.Line = a.Line) and (b.Char < a.Char))
end;

class operator TBufferCoord.GreaterThanOrEqual(a, b: TBufferCoord): Boolean;
begin
  Result :=  (b.Line < a.Line)
    or ((b.Line = a.Line) and (b.Char <= a.Char))
end;

class operator TBufferCoord.LessThan(a, b: TBufferCoord): Boolean;
begin
  Result :=  (b.Line > a.Line)
    or ((b.Line = a.Line) and (b.Char > a.Char))
end;

class operator TBufferCoord.LessThanOrEqual(a, b: TBufferCoord): Boolean;
begin
  Result :=  (b.Line > a.Line)
    or ((b.Line = a.Line) and (b.Char >= a.Char))
end;

class function TBufferCoord.Max(a, b: TBufferCoord): TBufferCoord;
begin
  if (b.Line < a.Line)
    or ((b.Line = a.Line) and (b.Char < a.Char))
  then
    Result := a
  else
    Result := b;
end;

class function TBufferCoord.Min(a, b: TBufferCoord): TBufferCoord;
begin
  if (b.Line < a.Line)
    or ((b.Line = a.Line) and (b.Char < a.Char))
  then
    Result := b
  else
    Result := a;
end;

class operator TBufferCoord.NotEqual(a, b: TBufferCoord): Boolean;
begin
  Result := (a.Char <> b.Char) or (a.Line <> b.Line);
end;

var
  rlist : System.Generics.Collections.TList<TSynFoldRange>;
  r : TSynFoldRange;

initialization
  System.Generics.Collections.TList<System.Word>.Create.Free;
  System.Generics.Collections.TList<Integer>.Create.Free;
  System.Generics.Collections.TList<TLineFoldInfo>.Create.Free;
  System.Generics.Collections.TEnumerable<System.Word>.Create.Free;
  rlist := System.Generics.Collections.TList<TSynFoldRange>.Create;
  rlist.Add(r);
  rlist.Free;
  System.Generics.Collections.TObjectStack<TSynUndoItem>.Create.Free;

end.

File: SourceCppRun64/SynEditDelphiInstances.cpp

#include <vcl.h>
#pragma hdrstop

#include "SynEditDelphiInstances.h"

using namespace std;
using namespace d2c_system;
using namespace System;
using namespace System::Generics::Defaults;

namespace Syneditdelphiinstances
{

__fastcall TSynUndoItem::TSynUndoItem() {}
__fastcall TSynFoldRangesCmp::TSynFoldRangesCmp() {}
__fastcall TLineFoldInfoCmp::TLineFoldInfoCmp() {}



int __fastcall TSynFoldRangesCmp::Compare(const TSynFoldRange& Left, const TSynFoldRange& Right)
{
	int result = 0;
	result = Left.FromLine - Right.FromLine;
	return result;
}

int __fastcall TLineFoldInfoCmp::Compare(const TLineFoldInfo& Left, const TLineFoldInfo& Right)
{
	int result = 0;
	result = Left.Line - Right.Line;
	return result;
}

System::Generics::Collections::TList__1<TSynFoldRange>* __fastcall CreateFoldRangeList()
{
	System::Generics::Collections::TList__1<TSynFoldRange>* result = nullptr;
	result = new System::Generics::Collections::TList__1<TSynFoldRange>(new TSynFoldRangesCmp());
	return result;
}

System::Generics::Collections::TList__1<TLineFoldInfo>* __fastcall CreateFoldInfoList()
{
	System::Generics::Collections::TList__1<TLineFoldInfo>* result = nullptr;
	result = new System::Generics::Collections::TList__1<TLineFoldInfo>(new TLineFoldInfoCmp());
	return result;
}


/* TSynEditFoldRange */

__fastcall TSynFoldRange::TSynFoldRange(int AFromLine, int AToLine/*# = -1*/, int AFoldType/*# = 1*/, int AIndent/*# = -1*/, bool ACollapsed/*# = false*/)
 : FromLine(AFromLine),
			ToLine(AToLine),
			FoldType(AFoldType),
			Indent(AIndent),
			Collapsed(ACollapsed)
{
}

int __fastcall TSynFoldRange::GetLinesCollapsed()
{
	int result = 0;
	if(Collapsed)
		result = ToLine - FromLine;
	else
		result = 0;
	return result;
}

void __fastcall TSynFoldRange::Move(int Count)
{
	FromLine += Count;
	ToLine += Count;
}

__fastcall TLineFoldInfo::TLineFoldInfo(int ALine, TFoldOpenClose AFoldOpenClose/*# = focOpen*/, int AFoldType/*# = 1*/, int AIndent/*# = -1*/)
 : Line(ALine),
			FoldOpenClose(AFoldOpenClose),
			FoldType(AFoldType),
			Indent(AIndent)
{
}

void __fastcall AddToFoldRangeList(System::Generics::Collections::TList__1<TSynFoldRange>* AList, const TSynFoldRange& cARange)
{
	TSynFoldRange ARange = cARange;
	AList->Add(ARange);
}

void __fastcall InsertIntoFoldRangeList(System::Generics::Collections::TList__1<TSynFoldRange>* AList, int AIndex, const TSynFoldRange& cARange)
{
	TSynFoldRange ARange = cARange;
	AList->Insert(AIndex, ARange);
}

bool __fastcall BinarySearchInFoldRangeList(System::Generics::Collections::TList__1<TSynFoldRange>* AList, const TSynFoldRange& cARange, int& AIndex)
{
	TSynFoldRange ARange = cARange;
	bool result = false;
	result = AList->BinarySearch(ARange, AIndex);
	return result;
}

// for SynEditCodeFolding

void __fastcall InsertIntoLineFoldList(System::Generics::Collections::TList__1<TLineFoldInfo>* AList, int AIndex, const TLineFoldInfo& cAInfo)
{
	TLineFoldInfo AInfo = cAInfo;
	AList->Insert(AIndex, AInfo);
}

bool __fastcall BinarySearchInLineFoldList(System::Generics::Collections::TList__1<TLineFoldInfo>* AList, const TLineFoldInfo& cAInfo, int& AIndex)
{
	TLineFoldInfo AInfo = cAInfo;
	bool result = false;
	result = AList->BinarySearch(AInfo, AIndex);
	return result;
}

/* TBufferCoord */

bool __fastcall operator == (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (A.Char == B.Char) && (A.Line == B.Line);
	return result;
}

bool __fastcall operator > (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (B.Line < A.Line) || ((B.Line == A.Line) && (B.Char < A.Char));
	return result;
}

bool __fastcall operator >= (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (B.Line < A.Line) || ((B.Line == A.Line) && (B.Char <= A.Char));
	return result;
}

bool __fastcall operator < (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (B.Line > A.Line) || ((B.Line == A.Line) && (B.Char > A.Char));
	return result;
}

bool __fastcall operator <= (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (B.Line > A.Line) || ((B.Line == A.Line) && (B.Char >= A.Char));
	return result;
}

/*#static*/
TBufferCoord __fastcall TBufferCoord::Max(const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	TBufferCoord result = {};
	if((B.Line < A.Line) || ((B.Line == A.Line) && (B.Char < A.Char)))
		result = A;
	else
		result = B;
	return result;
}

/*#static*/
TBufferCoord __fastcall TBufferCoord::Min(const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	TBufferCoord result = {};
	if((B.Line < A.Line) || ((B.Line == A.Line) && (B.Char < A.Char)))
		result = B;
	else
		result = A;
	return result;
}

bool __fastcall operator != (const TBufferCoord& cA, const TBufferCoord& cB)
{
	TBufferCoord A = cA;
	TBufferCoord B = cB;
	bool result = false;
	result = (A.Char != B.Char) || (A.Line != B.Line);
	return result;
}
System::Generics::Collections::TList__1<TSynFoldRange>* rlist = nullptr;
TSynFoldRange R = {};
static bool SynEditDelphiInstances_Initialized = false;

void SynEditDelphiInstances_initialization()
{
	if(SynEditDelphiInstances_Initialized)
		return;
	
	SynEditDelphiInstances_Initialized = true;
	
	delete new System::Generics::Collections::TList__1<System::Word>();
	delete new System::Generics::Collections::TList__1<int>();
	delete new System::Generics::Collections::TList__1<TLineFoldInfo>();
	delete new System::Generics::Collections::TEnumerable__1<System::Word>();
	rlist = new System::Generics::Collections::TList__1<TSynFoldRange>();
	rlist->Add(R);
	delete rlist;
	delete new System::Generics::Collections::TObjectStack__1<TSynUndoItem*>();
}
// using unit initialization order file, so unit singleton has not been created


}  // namespace SynEditDelphiInstances

